Skip interpreters with different version

dcermak opened the pull request:
TODO:

- [x] Write new tests or update the old ones to cover new functionality.
- [x] Update doc-strings where appropriate.
- [-] Update or write new documentation in `packit/packit.dev`.

<!-- notes for reviewers -->

Fixes #21

<!-- release notes footer -->

RELEASE NOTES BEGIN

`rpm-shim` now prefers the current interpreter for searching for the `rpm` module

RELEASE NOTES END

dcermak committed:
Add a regression test for https://github.com/packit/rpm-shim/issues/21

What we were observing is an ABI issue, python 3.11 would load the python3.6-rpm
module, leading to erratic behavior, like rpm.spec not existing. By trying to
call that module, we check whether such a missmatch is present or not.

mfocko commented on rpm/__init__.py:
```diff
@@ -71,6 +71,18 @@ def get_sitepackages_and_suffixes(interpreter):
+            .removeprefix("Python ")
+            .split(".")
+        )
+        if i_major != majorver or i_minor != minorver:
```
```suggestion
        if (i_major, i_minor) != (majorver, minorver):
```

I'd probably prefer this, since it's a bit easier to read.

gotmax23 commented on rpm/__init__.py:
```diff
@@ -71,6 +71,18 @@ def get_sitepackages_and_suffixes(interpreter):
             continue
+        # skip interpreters with a different major.minor version than the
+        # running interpreter
+        # importing their rpm module has most likely catastrophic consequences
```
Would it make sense to reorder the interpreters above and try `python{majorver}.{minorver}` first instead?. Since https://github.com/rpm-software-management/rpm/issues/2345, RPM python bindings are built against the Python 3.7+ stable ABI, so this change could break things that worked before, but it also would prevent segfaults and other nastiness in cases where they aren't.

On Fedora 40 with a Python 3.12 system interpreter, this seems to work:

``` console
$ uvx --from rpm --python 3.7 python -c 'import rpm; print(repr(rpm)); print(rpm.expandMacro("%_bindir"))'
<module 'rpm' from '/usr/lib64/python3.12/site-packages/rpm/__init__.py'>
/usr/bin
```

nforro commented on rpm/__init__.py:
```diff
@@ -64,16 +65,26 @@ def get_sitepackages_and_suffixes(interpreter):
+        # skip interpreters with a different major.minor version than the
+        # running interpreter
+        # importing their rpm module has most likely catastrophic consequences
+        i_major, i_minor, _ = re.sub(
+            r"^Python ",
+            "",
+            subprocess.check_output([interpreter, "--version"]).decode().strip(),
+        ).split(".")
```
Wouldn't it be better to call `platform.python_version_tuple()` in the script in `get_system_sitepackages_and_suffixes()` (and rename the function to something like `get_interpreter_info()`)?

dcermak replied:
Done

nforro commented on rpm/__init__.py:
```diff
@@ -64,16 +65,26 @@ def get_sitepackages_and_suffixes(interpreter):
         sitepackages_and_suffixes = get_sitepackages_and_suffixes(interpreter)
         logger.debug(
-            f"Collected sitepackages and extension suffixes for {interpreter}:\n"
+            f"Collected sitepackages and extension suffixes for {interpreter}:\n"  # noqa: E231
```
Why the `noqa` here and below?

dcermak replied:
flake8 started to report false positives here, on this line it was complaining about missing whitespace around `:` and in the other it was about that one should use `not in` for checks. But both are logged strings, so these warnings can be supressed (they were failing pre-commit)

nforro replied:
flake8 doesn't report anything for me when run manually nor during pre-commit. It makes no sense for these to be errors anyway, quick Google search suggests it was a [bug in flake8 that has been fixed in 6.1.0](https://stackoverflow.com/questions/77401175/how-to-make-flake8-ignore-syntax-within-strings/77402153).

dcermak replied:
The problem is that you have an old version of flake8 pinned in the pre-commit config, I have pushed a version bump that should fix this :crossed_fingers:

nforro commented on rpm/__init__.py:
```diff
@@ -64,16 +65,26 @@ def get_sitepackages_and_suffixes(interpreter):
     for interpreter in interpreters:
         if not Path(interpreter).is_file():
             continue
+        # skip interpreters with a different major.minor version than the
+        # running interpreter
+        # importing their rpm module has most likely catastrophic consequences
```
As was already established, since Python 3.7 a different minor version shouldn't have any consequences, let alone catastrophic.

dcermak replied:
That is not true for python 3.6 where we observed the bug though, there `rpm` does not contain the class/module `spec` and any operations involving the `specfile` module then fail with import errors.

nforro replied:
Yes, but the comment is still misleading.

dcermak replied:
Better now?

dcermak committed:
[pre-commit] Bump flake8 version & s/push/pre-push/

These can't be two commits because pre-commit forbids creating a commit with
unstaged changes to pre-commit-config.yaml

nforro commented on rpm/__init__.py:
```diff
@@ -64,19 +66,27 @@ def get_sitepackages_and_suffixes(interpreter):
     interpreters = [
         "/usr/libexec/platform-python",
-        f"/usr/bin/python{majorver}",
         f"/usr/bin/python{majorver}.{minorver}",
+        f"/usr/bin/python{majorver}",
```
Isn't the order change enough to solve your issue? Skipping breaks the ubi9 test case (3.9 rpm module with 3.11 interpreter).

dcermak replied:
Yes, it is

nforro commented on tests/import.py:
```diff
@@ -13,6 +13,29 @@ def test():
     # sanity check
     print("RPM conf dir:", rpm.expandMacro("%getconfdir"))
 
+    import site
+    import sys
+    from pathlib import Path
+
+    assert len(rpm.__path__) == 1
+    rpm_path = rpm.__path__[0]
+    search_paths = sys.path + site.getsitepackages()
+
+    for site_dir in search_paths:
+        try:
+            # is_relative_to is only present in python 3.9+
+            # relative_to raises a ValueError if the paths are not relative
+            # => retry
+            # if no exception is raised => found a match and break out
+            Path(rpm_path).relative_to(Path(site_dir))
+            break
+        except ValueError:
+            pass
+    else:
+        raise RuntimeError(
+            f"{rpm_path} is not relative to any python search path: {search_paths}"
+        )
+
```
What's the point of this check? It breaks the ubuntu test case.

dcermak replied:
I wanted to verify with this, that the imported rpm module belongs to the same interpreter. But the whole test is moot, if it is expected that python 3.11 uses the python 3.9 rpm module on CentOS/RHEL 9. I'll have to rework the test to correctly blow up on SLES and pass on RHEL

dcermak replied:
Turns out that just trying to instantiate the `rpm.spec` module/class was enough for a regression test

dcermak committed:
Switch the search order of interpreters so that we use the $major.$minor version first

Currently we can run into the problem, that we have two interpreters installed,
both have the rpm module, but only the interpreter with the same version is
desirable (e.g. python 3.11 & python3.6 on openSUSE Leap 15). We can easily fix
this issue by first checking the $major.$minor version and then the "other"
python version

dcermak committed:
[ci] Specify the full url for openSUSE Leap on registry.o.o

nforro reviewed (approved):
Great, thanks!

nforro commented:
@dcermak Could you please just edit the release notes? I would then merge this and do a new release.

dcermak commented:
> @dcermak Could you please just edit the release notes? I would then merge this and do a new release.

Done

nforro commented:
> > @dcermak Could you please just edit the release notes? I would then merge this and do a new release.
> 
> Done

I meant in the PR description, changelog is handled by release automation.

dcermak commented:
> > > @dcermak Could you please just edit the release notes? I would then merge this and do a new release.
> > 
> > 
> > Done
> 
> I meant in the PR description, changelog is handled by release automation.

oops, fixed that
